api/server/errorhandler.go

```diff
@@ -17,18 +21,80 @@ func makeErrorHandler(err error) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
 		statusCode := httpstatus.FromError(err)
 		vars := mux.Vars(r)
-		if apiVersionSupportsJSONErrors(vars["version"]) {
-			response := &types.ErrorResponse{
-				Message: err.Error(),
-			}
-			_ = httputils.WriteJSON(w, statusCode, response)
-		} else {
+		version := vars["version"]
+
+		const firstAPIVersionWithJSONErrors = "1.24"
+		if version != "" && versions.LessThan(version, firstAPIVersionWithJSONErrors) {
 			http.Error(w, status.Convert(err).Message(), statusCode)
+			return
+		}
+
+		response := marshalErrorResponse(err)
+
+		const firstAPIVersionWithJSONMultiErrors = "1.44"
+		if version != "" && versions.LessThan(version, firstAPIVersionWithJSONMultiErrors) {
+			response.Errors = nil
 		}
+
+		_ = httputils.WriteJSON(w, statusCode, response)
 	}
 }
 
-func apiVersionSupportsJSONErrors(version string) bool {
-	const firstAPIVersionWithJSONErrors = "1.23"
-	return version == "" || versions.GreaterThan(version, firstAPIVersionWithJSONErrors)
+type joiningError interface {
+	Unwrap() []error
+	Error() string
+}
+
+type wrappingError interface {
+	Unwrap() error
+	Error() string
+}
+
+// marshalErrorResponse returns a tree of [*types.ErrorResponse] representing the argument err.
+func marshalErrorResponse(err error) *types.ErrorResponse {
+	if errdefs.IsHTTPOnlyError(err) {
+		if wrappedErr := errors.Unwrap(err); wrappedErr != nil {
+			err = wrappedErr
+		}
+	}
+
+	if err, ok := err.(joiningError); ok {
+		return marshalJoinedErrors(err)
+	}
+
+	if err, ok := err.(wrappingError); ok {
+		return &types.ErrorResponse{
+			Message: err.Error(),
+			Errors:  []*types.ErrorResponse{marshalErrorResponse(err.Unwrap())},
+		}
+	}
+
+	return &types.ErrorResponse{
+		Message: err.Error(),
+	}
+}
+
+func marshalJoinedErrors(err joiningError) *types.ErrorResponse {
+	var subErrors []*types.ErrorResponse
+	for _, subErr := range err.Unwrap() {
+		if subErr != nil {
+			subErrors = append(subErrors, marshalErrorResponse(subErr))
+		}
+	}
+
+	errMsg := err.Error()
+	if len(subErrors) > 1 {
+		var stringErrs []string
+		for _, subErr := range subErrors {
+			stringErrs = append(stringErrs, strings.Replace(subErr.Message, "\n", "\n\t", -1))
+		}
+		errMsg = fmt.Sprintf("%d errors occurred:\n\t* %s", len(subErrors), strings.Join(stringErrs, "\n\t* "))
+	} else {
+		subErrors = nil
+	}
```

Comment:
I see what you're getting at here. It makes sense to flatten an `errors.Join()` error which unwraps to a single error as the outer error carries no information and simply passes through the inner error string verbatim. However, I see two problems with how you're going about it.

First, recursion. The singular sub-error might itself be a `joiningError` or `wrappingError`. Marshaling the outer `joiningError` and discarding the marshaled sub-error also truncates that branch of the error tree. Flatten by discarding the useless wrapper, not its contents.

Second, it's not safe to assume that all error types which satisfy `joiningError` have the same (lack of) semantics. We can't just test directly for whether the error is an `errors.Join()` error as the concrete type is not exported, so I propose dealing with it heuristically. Flatten the error if and only if no information would be lost by discarding it.

```suggestion
	} else if len(subErrors) == 1 && errMsg == subErrors[0].Message {
		return subErrors[0]
	}
```